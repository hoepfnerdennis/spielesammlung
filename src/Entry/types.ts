export type EntryProps = {
  name: string;
  description: string;
  players: {
    from: string;
    to: string;
  };
  age: string;
  duration: string;
};
