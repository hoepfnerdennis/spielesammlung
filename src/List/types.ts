interface IBaseGame {
  name: string;
  description: string;
  playersFrom: number;
  playersTo: number;
  age: string;
  duration: string;
  favorite: boolean;
  simpleRules: boolean;
}

export interface IGame extends IBaseGame {
  image: string;
}

enum LinkTypes {
  'Asset',
  'Space',
  'Environment',
  'ContentType',
}

enum SysTypes {
  'Link',
  'Asset',
}

interface IAPIGame extends IBaseGame {
  image?: {
    sys: {
      type: string;
      linkType: LinkTypes;
      id: string;
    };
  };
}

export interface IAsset {
  sys: {
    space: {
      sys: {
        type: string;
        linkType: LinkTypes;
        id: string;
      };
    };
    id: string;
    type: SysTypes;
    createdAt: string;
    updatedAt: string;
    environment: {
      sys: {
        id: string;
        type: SysTypes;
        linkType: LinkTypes;
      };
    };
    revision: number;
    locale: string;
  };
  fields: {
    title: string;
    description: string;
    file: {
      url: string;
      details: {
        size: number;
        image: {
          width: number;
          height: number;
        };
      };
      fileName: string;
      contentType: string;
    };
  };
}

export interface IAPIResponse {
  sys: {
    type: string;
  };
  total: number;
  skip: number;
  limit: number;
  items: [
    {
      sys: {
        space: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        id: string;
        type: string;
        createdAt: string;
        updatedAt: string;
        environment: {
          sys: {
            id: string;
            type: string;
            linkType: LinkTypes;
          };
        };
        revision: number;
        contentType: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        locale: string;
      };
      fields: IAPIGame;
    }
  ];
  includes?: {
    Asset: IAsset[];
  };
}

/*
{
  "sys": {
    "type": "Array"
  },
  "total": 3,
  "skip": 0,
  "limit": 100,
  
  "includes": {
    "Asset": [
      {
        "sys": {
          "space": {
            "sys": {
              "type": "Link",
              "linkType": "Space",
              "id": "9sxha2f3gm24"
            }
          },
          "id": "2Txvyka47nhfVoh2Sc1OUh",
          "type": "Asset",
          "createdAt": "2020-03-20T14:04:02.201Z",
          "updatedAt": "2020-03-20T14:04:02.201Z",
          "environment": {
            "sys": {
              "id": "master",
              "type": "Link",
              "linkType": "Environment"
            }
          },
          "revision": 1,
          "locale": "en-US"
        },
        "fields": {
          "title": "Mensch ärgere dich nicht",
          "description": "Mensch ärgere dich nicht",
          "file": {
            "url": "//images.ctfassets.net/9sxha2f3gm24/2Txvyka47nhfVoh2Sc1OUh/d277e6dc09f15b5bb5a134ff7082fe4a/photo-1582921017967-79d1cb6702ee",
            "details": {
              "size": 192767,
              "image": {
                "width": 2100,
                "height": 1400
              }
            },
            "fileName": "photo-1582921017967-79d1cb6702ee",
            "contentType": "image/jpeg"
          }
        }
      }
    ]
  }
}
 */
