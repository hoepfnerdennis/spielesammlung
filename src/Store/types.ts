interface IBaseGame {
  name: string;
  description: string;
  playersFrom: number;
  playersTo: number;
  age: string;
  duration: string;
  favorite: boolean;
  simpleRules: boolean;
}

export interface IGame extends IBaseGame {
  image: string;
}

enum LinkTypes {
  'Asset',
  'Space',
  'Environment',
  'ContentType',
}

enum SysTypes {
  'Link',
  'Asset',
}

interface IAPIGame extends IBaseGame {
  image?: {
    sys: {
      type: string;
      linkType: LinkTypes;
      id: string;
    };
  };
}

export interface IAsset {
  sys: {
    space: {
      sys: {
        type: string;
        linkType: LinkTypes;
        id: string;
      };
    };
    id: string;
    type: SysTypes;
    createdAt: string;
    updatedAt: string;
    environment: {
      sys: {
        id: string;
        type: SysTypes;
        linkType: LinkTypes;
      };
    };
    revision: number;
    locale: string;
  };
  fields: {
    title: string;
    description: string;
    file: {
      url: string;
      details: {
        size: number;
        image: {
          width: number;
          height: number;
        };
      };
      fileName: string;
      contentType: string;
    };
  };
}

export interface IAPIResponse {
  sys: {
    type: string;
  };
  total: number;
  skip: number;
  limit: number;
  items: [
    {
      sys: {
        space: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        id: string;
        type: string;
        createdAt: string;
        updatedAt: string;
        environment: {
          sys: {
            id: string;
            type: string;
            linkType: LinkTypes;
          };
        };
        revision: number;
        contentType: {
          sys: {
            type: string;
            linkType: LinkTypes;
            id: string;
          };
        };
        locale: string;
      };
      fields: IAPIGame;
    }
  ];
  includes?: {
    Asset: IAsset[];
  };
}

export interface IFilters {
  playersFrom: number | undefined;
  playersTo: number | undefined;
  favorite: boolean | undefined;
  search: string | undefined;
}

export interface IState {
  games: Array<IGame>;
  filters: IFilters;
}

export type Reducer = (state: IState, action: IAction) => IState;

export type Dispatch = React.Dispatch<IAction>;

export interface IAction {
  type: string;
  payload: any;
}
