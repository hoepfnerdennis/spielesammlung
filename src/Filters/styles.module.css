.filters {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;

    max-width: 800px;
    margin: 1rem auto;
}

.element {
    margin-top: 1rem;
}

@media (min-width: 900px) {
    .filters {
        flex-direction: row;
    }
    .element {
        margin-top: 0;
    }
}