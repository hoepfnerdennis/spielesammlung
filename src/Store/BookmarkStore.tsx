import { useEffect } from 'react';
import makeStore from '../utils/makeStore';
import useLocalStorage from '../utils/useLocalStoreage';

interface IAction {
  type: string;
  payload: any;
}

export enum ActionTypes {
  'SET_BOOKMARKS' = 'SET_BOOKMARKS',
  'ADD_BOOKMARK' = 'ADD_BOOKMARK',
  'REMOVE_BOOKMARK' = 'REMOVE_BOOKMARK',
}

const initialState: string[] = localStorage.getItem('bookmarks')?.split(',') || [];

const dataReducer = (state: string[], action: IAction): string[] => {
  switch (action.type) {
    case ActionTypes.SET_BOOKMARKS:
      return action.payload;
    case ActionTypes.ADD_BOOKMARK:
      return [...state, action.payload];
    case ActionTypes.REMOVE_BOOKMARK:
      return state.filter((value) => value !== action.payload);
    default:
      return state;
  }
};

const [BookmarksProvider, useBookmarks, useDispatch] = makeStore<string[], IAction>(
  dataReducer,
  initialState
);

const useMark = (): ((id: string) => void) => {
  const dispatch = useDispatch();
  const bookmarks = useBookmarks();
  const { setItem } = useLocalStorage();

  const mark = (id: string): void => {
    if (bookmarks.includes(id)) {
      return dispatch({ type: ActionTypes.REMOVE_BOOKMARK, payload: id });
    }
    return dispatch({ type: ActionTypes.ADD_BOOKMARK, payload: id });
  };

  useEffect(() => {
    setItem<string[]>('bookmarks', bookmarks);
  }, [bookmarks, setItem]);

  return mark;
};

export { BookmarksProvider, useBookmarks, useMark };
