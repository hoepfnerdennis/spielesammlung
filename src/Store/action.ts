import { useCallback, useState, useEffect } from 'react';
import {
  IGame,
  IAPIResponse,
  FiltersConfig,
  FilterKey,
  FilterValue,
  SetFilterFunc,
  ActiveFiltersMap,
} from './types';
import { mapResultsToGames } from '../utils';

const SPACE_ID = '9sxha2f3gm24';
const API_TOKEN = '7LDIC95TsrYOfZwEQnbAuMHtij97kfk5r1dIRiGqT8M';
// For local testing replace BASE_URI with
// const BASE_URI = 'http://localhost:3001/response.json';
const BASE_URI = `https://cdn.contentful.com/spaces/${SPACE_ID}/environments/master/entries?access_token=${API_TOKEN}`;

const filtersConfig: FiltersConfig = {
  playersFrom: '[lte]',
  playersTo: '[gte]',
  name: '[match]',
  favorite: '',
};

const fetchGamesAction = async (filters: Map<FilterKey, FilterValue>): Promise<IGame[]> => {
  let url = `${BASE_URI}&content_type=game`;
  filters.forEach((value, key) => {
    url += `&fields.${FilterKey[key]}${filtersConfig[FilterKey[key]]}=${value}`; // &fields.playersTo[gte]=${from}`;
  });

  try {
    const response = await fetch(url);
    const data: IAPIResponse = await response.json();
    const gamesFromAPI: IGame[] = mapResultsToGames(data);
    return gamesFromAPI;
  } catch {
    return [];
  }
};

const useData = (): {
  games: IGame[];
  setFilter: SetFilterFunc;
  activeFilters: ActiveFiltersMap;
} => {
  const [games, setGames] = useState<IGame[]>([]);
  const [activeFilters, setActiveFilters] = useState<ActiveFiltersMap>(new Map());

  useEffect(() => {
    const fetchGames = async (): Promise<void> => {
      const newGames = await fetchGamesAction(activeFilters);
      setGames(newGames);
    };
    fetchGames();
  }, [activeFilters]);

  const addFilter = useCallback((key: FilterKey, value: FilterValue): void => {
    setActiveFilters(activeFiltersMap => {
      if (activeFiltersMap.get(key) === value) {
        return activeFiltersMap;
      }
      const newMap = new Map(activeFiltersMap);
      newMap.set(key, value);
      return newMap;
    });
  }, []);

  const removeFilter = useCallback((key: FilterKey): void => {
    setActiveFilters(activeFiltersMap => {
      if (!activeFiltersMap.has(key)) {
        return activeFiltersMap;
      }
      const newMap = new Map(activeFiltersMap);
      newMap.delete(key);
      return newMap;
    });
  }, []);

  const setFilter = useCallback(
    (key: FilterKey) => (value: FilterValue | undefined): void => {
      if (value) {
        addFilter(key, value);
      } else {
        removeFilter(key);
      }
    },
    [addFilter, removeFilter]
  );

  return { games, setFilter, activeFilters };
};

export default useData;
