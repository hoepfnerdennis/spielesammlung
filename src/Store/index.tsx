import React, { createContext, useReducer } from 'react';
import { IState, Reducer, IAction } from './types';

export const FETCH_DATA_FINISHED = 'FETCH_DATA_FINISHED';
export const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
export const SET_PALYER_FROM = 'SET_PALYER_FROM';
export const SET_PLAYER_TO = 'SET_PLAYER_TO';
export const SET_FAVORITE = 'SET_FAVORITE';

const initialState: IState = {
  games: [],
  filters: {
    playersFrom: undefined,
    playersTo: undefined,
    favorite: undefined,
    search: undefined,
  },
};

export const Store = createContext<IState | any>(initialState);

const reducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case FETCH_DATA_FINISHED:
      return { ...state, games: action.payload };

    case SET_PALYER_FROM:
      return { ...state, filters: { ...state.filters, playersFrom: action.payload } };

    case SET_PLAYER_TO:
      return { ...state, filters: { ...state.filters, playersTo: action.payload } };

    case SET_SEARCH_TERM:
      return { ...state, filters: { ...state.filters, search: action.payload } };

    case SET_FAVORITE:
      return { ...state, filters: { ...state.filters, favorite: action.payload } };

    default:
      return state;
  }
};

export const StoreProvider: React.SFC = ({ children }): JSX.Element => {
  const [state, dispatch] = useReducer<Reducer>(reducer, initialState);
  return <Store.Provider value={{ state, dispatch }}>{children}</Store.Provider>;
};
