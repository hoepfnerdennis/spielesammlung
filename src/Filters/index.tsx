import React, { useMemo, useState } from 'react';
import styles from './styles.module.css';
import Select from '../Select';
import Checkbox from '../Checkbox';
import { order } from '../utils';
import Layer from '../Layer';
import { IGame, ActiveFiltersMap, FilterKey, SetFilterFunc } from '../Store/types';

const Filters: React.FC<{
  games: IGame[];
  setFilter: SetFilterFunc;
  activeFilters: ActiveFiltersMap;
}> = ({ games, setFilter, activeFilters }): JSX.Element => {
  const [displayFilterLayer, setDisplayFilterLayer] = useState<boolean>(false);

  const playersFromValues: string[] = useMemo(
    () =>
      games
        .map(game => game.playersFrom)
        .filter((players, index, array) => array.indexOf(players) === index)
        .sort(order)
        .map(f => f.toString()),
    [games]
  );

  const playersToValues: string[] = useMemo(
    () =>
      games
        .map(game => game.playersTo)
        .filter((players, index, array) => array.indexOf(players) === index)
        .sort(order)
        .map(f => f.toString()),
    [games]
  );

  return (
    <>
      <button
        type="button"
        className={styles.button}
        onClick={(): void => setDisplayFilterLayer(true)}>
        Filter anzeigen
      </button>
      {displayFilterLayer && (
        <Layer closeLayer={(): void => setDisplayFilterLayer(false)}>
          <h2 className={styles.headline}>Filtere die Spiele</h2>
          <div className={styles.element}>
            <Select
              values={playersFromValues}
              value={activeFilters.get(FilterKey.playersFrom) || ''}
              onChange={setFilter(FilterKey.playersFrom)}
              label="Nur Spiele ab"
              valueSuffix="Spieler"
            />
          </div>
          <div className={styles.element}>
            <Select
              values={playersToValues}
              value={activeFilters.get(FilterKey.playersTo) || ''}
              onChange={setFilter(FilterKey.playersTo)}
              label="Nur Spiele bis"
              valueSuffix="Spieler"
            />
          </div>
          <div className={styles.element}>
            <Checkbox
              checked={activeFilters.has(FilterKey.favorite)}
              label={
                activeFilters.get(FilterKey.favorite)
                  ? 'Alle Spiele anzeigen'
                  : 'Nur Empfehlungen anzeigen'
              }
              onChange={setFilter(FilterKey.favorite)}
            />
          </div>
        </Layer>
      )}
    </>
  );
};

export default Filters;
