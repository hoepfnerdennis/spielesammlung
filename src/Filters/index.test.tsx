import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Filters from './index';
import { ActiveFiltersMap, FilterKey } from '../Store/types';
import getGame from '../../mockData/test-games';

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type,@typescript-eslint/no-explicit-any
jest.mock('../Layer/index', () => ({ children, closeLayer }: any) => (
  <button onClick={closeLayer} data-testid="layer" type="button">
    {children}
  </button>
));

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
jest.mock('../Select', () => () => 'Select');

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
jest.mock('../Checkbox', () => () => 'Checkbox');

describe('Filters', () => {
  it.each([
    ['no active filter', new Map()],
    ['favorite selected', new Map().set(FilterKey.favorite, 'true')],
  ])('should render %s', (_, activeFilters) => {
    const games = [getGame(), getGame(), getGame(), getGame()];
    const setFilter = jest.fn();
    const { container, getByText } = render(
      <Filters activeFilters={activeFilters} games={games} setFilter={setFilter} />
    );
    fireEvent.click(getByText('Filter anzeigen'));
    expect(container).toMatchSnapshot();
  });
  it('should open and close layer', () => {
    const activeFilters: ActiveFiltersMap = new Map();
    const games = [getGame(), getGame(), getGame(), getGame()];
    const setFilter = jest.fn();
    const { container, getByText, getByTestId } = render(
      <Filters activeFilters={activeFilters} games={games} setFilter={setFilter} />
    );
    fireEvent.click(getByText('Filter anzeigen'));
    expect(container).toMatchSnapshot();

    fireEvent.click(getByTestId('layer'));
    expect(container).toMatchSnapshot();
  });
});
