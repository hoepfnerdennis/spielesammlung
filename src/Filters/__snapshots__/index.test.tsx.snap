// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters should open and close layer 1`] = `
<div>
  <button
    class="button"
    type="button"
  >
    Filter anzeigen
  </button>
  <button
    data-testid="layer"
    type="button"
  >
    <h2
      class="headline"
    >
      Filtere die Spiele
    </h2>
    <div
      class="element"
    >
      Select
    </div>
    <div
      class="element"
    >
      Select
    </div>
    <div
      class="element"
    >
      Checkbox
    </div>
  </button>
</div>
`;

exports[`Filters should open and close layer 2`] = `
<div>
  <button
    class="button"
    type="button"
  >
    Filter anzeigen
  </button>
</div>
`;

exports[`Filters should render favorite selected 1`] = `
<div>
  <button
    class="button"
    type="button"
  >
    Filter anzeigen
  </button>
  <button
    data-testid="layer"
    type="button"
  >
    <h2
      class="headline"
    >
      Filtere die Spiele
    </h2>
    <div
      class="element"
    >
      Select
    </div>
    <div
      class="element"
    >
      Select
    </div>
    <div
      class="element"
    >
      Checkbox
    </div>
  </button>
</div>
`;

exports[`Filters should render no active filter 1`] = `
<div>
  <button
    class="button"
    type="button"
  >
    Filter anzeigen
  </button>
  <button
    data-testid="layer"
    type="button"
  >
    <h2
      class="headline"
    >
      Filtere die Spiele
    </h2>
    <div
      class="element"
    >
      Select
    </div>
    <div
      class="element"
    >
      Select
    </div>
    <div
      class="element"
    >
      Checkbox
    </div>
  </button>
</div>
`;
